### PR TITLE
Track log_model APIs

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/16611?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16611/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16611/merge#subdirectory=skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16611/merge
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Track log_model APIs and add tests
Depends on https://github.com/mlflow/mlflow/pull/16549

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/transformers/test_transformers_model_export.py [hunk 1]
```diff
@@ -3882,3 +3885,33 @@ def test_log_model_skip_validating_serving_input_for_local_checkpoint(
         mock_validate_input.assert_not_called()
     else:
         mock_validate_input.assert_called_once()
+
+
+def test_log_model_sends_telemetry_record(mock_requests, base_model):
```
Comment (harupy) [inline tests/transformers/test_transformers_model_export.py]: do we need to test all flavors? Can we pick one and test it in `tests/telemetry`?

Comment (serena-ruan) [reply tests/transformers/test_transformers_model_export.py]: Based on my previous tests, it's easy to miss edge cases :( The test itself is not complex so I just add for all flavors

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -1935,6 +1935,30 @@ def _mock_dbutils(globals_dict):
             del globals_dict[module_name]
 
 
+@contextmanager
```
Comment (serena-ruan) [inline mlflow/models/utils.py]: This is one of the main changes

File: mlflow/telemetry/parser.py [hunk 3]
```diff
@@ -44,16 +44,12 @@ class LogModelParser(TelemetryParser):
     def extract_params(
         cls, func: Callable[..., Any], arguments: dict[str, Any]
     ) -> Optional[LogModelParams]:
-        if flavor := _parse_flavor_from_module_name(func.__module__):
-            flavor = flavor
-        else:
+        flavor = _parse_flavor_from_module_name(func.__module__)
```
Comment (serena-ruan) [inline mlflow/telemetry/parser.py:47]: Another main change

File: mlflow/telemetry/track.py [hunk 4]
```diff
@@ -39,6 +41,30 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
+@contextmanager
+def _avoid_telemetry_tracking():
```
Comment (serena-ruan) [inline mlflow/telemetry/track.py]: This is the main change --> this avoids two issues:
1. MLflow invokes databricks-agents APIs, which invokes MLflow APIs again. This check can avoid tracking internal APIs.
2. When we log model using code-based logging, the model file is executed directly and could cause more telemetry logging. This avoids that during model loading/logging.

File: mlflow/telemetry/track.py [hunk 5]
```diff
@@ -24,7 +25,8 @@ def wrapper(*args, **kwargs):
         success = True
         start_time = time.time()
         try:
-            return func(*args, **kwargs)
+            with _avoid_telemetry_tracking():
```
Comment (TomeHirata) [inline mlflow/telemetry/track.py]: Maybe add a comment why we have this logic (e.g, "Avoid tracking nested api calls from databricks-agents or log_model")

File: mlflow/telemetry/parser.py [hunk 6]
```diff
@@ -67,28 +63,39 @@ def extract_params(
...
+            if model is None and arguments.get("loader_module"):
+                return ModelType.LOADER_MODULE
+
+            try:
+                from mlflow.pyfunc.model import ResponsesAgent
+            except ImportError:
+                pass
+            else:
+                if isinstance(model, ResponsesAgent):
+                    return ModelType.RESPONSES_AGENT
```
Comment (TomeHirata) [inline mlflow/telemetry/parser.py:79]: nit: can't we include the else block inside try block since this code is executed when `ResponsesAgent` can be imported anyway?

**Review by TomeHirata — APPROVED** [review]
LGTM!

File: tests/transformers/test_transformers_model_export.py (hunk 1 above)
Comment (harupy) [reply tests/transformers/test_transformers_model_export.py]: Got it, let's keep them :)

File: mlflow/telemetry/track.py (hunk 4 above)
Comment (harupy) [reply mlflow/telemetry/track.py]: Can you add this ^ in the comment?

Comment (serena-ruan) [reply mlflow/telemetry/track.py]: Ok

File: mlflow/telemetry/track.py [hunk 7]
```diff
@@ -39,6 +41,26 @@ def wrapper(*args, **kwargs):
...
+    original_func = invoked_from_internal_api
+
+    def mock_invoked_from_internal_api():
+        return True
+
+    try:
+        invoked_from_internal_api = mock_invoked_from_internal_api
+        yield
+    finally:
+        invoked_from_internal_api = original_func
```
Comment (harupy) [inline mlflow/telemetry/track.py]: does this work fine in a multi-thread scenario?

File: tests/langgraph/test_langgraph_model_export.py [hunk 8]
```diff
@@ -77,3 +80,33 @@ def test_langgraph_model_invoke_with_dictionary_params(monkeypatch):
     assert len(pyfunc_model.predict(input_example, params)[0]["messages"]) == len(
         result["messages"]
     )
+
+
+def test_log_model_sends_telemetry_record(mock_requests):
+    """Test that log_model sends telemetry records."""
```
Comment (harupy) [inline tests/langgraph/test_langgraph_model_export.py]:
```suggestion
```

let's remove useless docstring.

File: tests/onnx/test_onnx_model_export.py [hunk 9]
```diff
@@ -735,3 +739,33 @@ def test_model_log_with_metadata(onnx_model):
...
+    get_telemetry_client().flush()
+
+    # Check that telemetry record was sent
+    assert len(mock_requests) == 1
+    record = mock_requests[0]
+    data = json.loads(record["data"])
+    assert data["api_module"] == mlflow.onnx.log_model.__module__
+    assert data["api_name"] == "log_model"
+    assert data["params"] == asdict(
+        LogModelParams(
```
Comment (harupy) [inline tests/onnx/test_onnx_model_export.py:760]: can we add a `asdict` method to `LogModelParams`?

File: mlflow/telemetry/track.py (hunk 7 above)
Comment (serena-ruan) [reply mlflow/telemetry/track.py]: It actually doesn't work well 🤔 I should add a contextVar for this

File: tests/onnx/test_onnx_model_export.py (hunk 9 above)
Comment (serena-ruan) [reply tests/onnx/test_onnx_model_export.py:760]: I don't think we need to, this is widely used just for tests.

File: mlflow/telemetry/track.py (hunk 7 above)
Comment (harupy) [reply mlflow/telemetry/track.py]: This look like a hack. Is this really the only option?

File: tests/onnx/test_onnx_model_export.py (hunk 9 above)
Comment (harupy) [reply tests/onnx/test_onnx_model_export.py:760]: Fair :)

File: mlflow/telemetry/track.py (hunk 7 above)
Comment (serena-ruan) [reply mlflow/telemetry/track.py]: I think it's quite tricky.. In general we need to avoid below scenarios:
1. The current API is invoked by another MLflow API --> this is solved by inspecting the current frames
2. The current API invokes a 3rd-party API (e.g. databricks-agents), and then it invokes MLflow APIs: e.g.mflow.genai.evaluate --> databricks-agents --> mlflow.trace. This cannot be solved by inspecting the current frames, but probably solvable by inspecting the calling stack which should cost more time
3. Direct API calls when using code-based logging: e.g. user's model enabled autolog, and when the model is loaded/logged, user's code is executed and cause this issue --> this is not solvable even by inspecting the calling stack

So the most straightforward way is to suppress logging more APIs when we decide to log the current API.
Do you have better ideas?

Comment (serena-ruan) [reply mlflow/telemetry/track.py]: cc @B-Step62 as we discussed part of this in the initial PR, and inspecting frames solution was chosen to solve scenario 1. While I found scenario 2 and 3 when decorating more APIs 😢

Comment (serena-ruan) [reply mlflow/telemetry/track.py]: Updated this to use a boolean directly, hope it looks better now @harupy

File: mlflow/telemetry/utils.py [hunk 10]
```diff
@@ -16,9 +18,31 @@ def is_telemetry_disabled() -> bool:
...
+# ContextVar to disable telemetry tracking in the current thread.
+# This is thread-local to avoid race conditions when multiple threads are running in parallel.
+_disable_telemetry_tracking_var = ContextVar("disable_telemetry_tracking", default=False)
+
+
+@contextmanager
+def _avoid_telemetry_tracking():
+    """
+    Context manager to disable telemetry tracking in the following scenarios:
+    1. Circular API calls: When MLflow invokes Databricks Agents APIs, which in turn call back
```
Comment (harupy) [inline mlflow/telemetry/utils.py]:
```suggestion
    1. Circular API calls: When MLflow invokes `databricks-agents` APIs, which in turn call back
```

File: mlflow/telemetry/utils.py [hunk 11]
```diff
@@ -16,9 +18,31 @@ def is_telemetry_disabled() -> bool:
...
         del frame
+
+
+# ContextVar to disable telemetry tracking in the current thread.
+# This is thread-local to avoid race conditions when multiple threads are running in parallel.
+_disable_telemetry_tracking_var = ContextVar("disable_telemetry_tracking", default=False)
+
+
+@contextmanager
+def _avoid_telemetry_tracking():
```
Comment (harupy) [inline mlflow/telemetry/utils.py]:
```suggestion
def _disable_telemetry():
```

can we rename this?

Comment (harupy) [inline mlflow/telemetry/utils.py]: Is this testable?

Comment (serena-ruan) [reply mlflow/telemetry/utils.py]: Yes I added test in test_utils.py

File: tests/telemetry/test_utils.py [hunk 12]
```diff
@@ -13,3 +20,38 @@ def test_is_telemetry_disabled(monkeypatch):
...
+    def thread_target(x):
+        if x % 2 == 0:
+            with _avoid_telemetry_tracking():
+                test_func()
+        else:
+            test_func()
+
+    with ThreadPoolExecutor(max_workers=10) as executor:
+        futures = [executor.submit(thread_target, i) for i in range(10)]
+        [f.result() for f in futures]
```
Comment (harupy) [inline tests/telemetry/test_utils.py]: Let's use for-loop. This looks like `forgot to assign the result`.

File: tests/telemetry/test_utils.py [hunk 13]
```diff
@@ -13,3 +20,38 @@ def test_is_telemetry_disabled(monkeypatch):
     with monkeypatch.context() as m:
         m.setenv("DO_NOT_TRACK", "true")
         assert is_telemetry_disabled() is True
+
+
+def test_avoid_telemetry_tracking(mock_requests):
+    @track_api_usage
+    def test_func():
+        pass
```
Comment (harupy) [inline tests/telemetry/test_utils.py:28]: Can we test a scenario like this?


```python
@track_api_usage
def test_func_1():
    test_func_2()


def test_func_2():
    test_func_3()


@track_api_usage
def test_func_3():
    pass


test_func_1()
```

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!
